### PR TITLE
Actionx parse

### PR DESCRIPTION
**Pull request by joakim-hove**
Will parse - and evaluate - the condition part of the ACTIONX keyword. Remaining steps:

1. Initialize the `ActionContext` object with proper values from the simulation, right now it is just explicitly assigned in tests.

2. Actually apply the keywords in the `ACTIONX` block.

3. Hook it all up in the simulator. 

Continuation of: https://github.com/OPM/opm-common/pull/539

### CODE REVIEW

**Comment by joakim-hove** [comment]
jenkins build this please

**Comment by joakim-hove** [comment]
jenkins build this please

**Comment by joakim-hove** [comment]
jenkins build this please

**Comment by joakim-hove** [comment]
jenkins build this please

**Comment by joakim-hove** [comment]
Review anyone?

**Comment by bska** [comment]
> Review anyone?

I can have a look, but I won't be able to do so before tonight (CET).  Anyone else feel free to jump in.  In the meantime we should verify that Jenkins still approves following the rebase.

**Comment by bska** [comment]
jenkins build this please

File: opm/parser/eclipse/EclipseState/Schedule/ActionAST.hpp [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+    double value(const ActionContext& context) const;
+    TokenType type;
+    void add_child(const ASTNode& child);
+    size_t size() const;
+
+private:
+    std::string func;
+    std::vector<std::string> arg_list;
+    double number;
+    std::vector<ASTNode> children;
```
Comment (bska) [inline opm/parser/eclipse/EclipseState/Schedule/ActionAST.hpp]: Do note that having a `vector<T>` as a member of `T` (in this case, `T = ASTNode`) does to a certain extent depend on undefined behaviour&mdash;at least until C++17 which explicitly blesses this particular usage.  It depends on being able to instantiate objects of type `std::allocator<T>` when `T` is an incomplete type and, subtly, on `vector<>` being implemented in terms of `T*`.

You may want to add a comment to that effect here.

Comment (joakim-hove) [reply opm/parser/eclipse/EclipseState/Schedule/ActionAST.hpp]: OK - I'll do that; I must admit these C++ peculiarities annoy me a bit. If it compiles without warnings (for some reasonably high warning level); and works - I would like to think the code was decent from a standards point of view.

Comment (bska) [reply opm/parser/eclipse/EclipseState/Schedule/ActionAST.hpp]: > I must admit these C++ peculiarities annoy me a bit. If it compiles without warnings (for some reasonably high warning level); and works - I would like to think the code was decent from a standards point of view.

I appreciate the sentiment.  To be fair, with libstdc++ the compiler only sees something that looks a lot like
```C++
class ASTNode
{
// ...
private:
    class V
    {
    // ...
    private:
        ASTNode* __x;
    // ...
    };
// ...
    V children;
};
```
and this is completely fine from a language perspective and therefore no diagnostic is required.

**Pull request merged**